Fix torch.distributed.breakpoint

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115656
* __->__ #115705

Switches from calling breakpoint() internally to using a subclass of
Pdb.

Fixes #115685

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @yf225

wconstab committed:
Fix torch.distributed.breakpoint

Switches from calling breakpoint() internally to using a subclass of
Pdb.

Fixes #115685

[ghstack-poisoned]

wanchaol commented on torch/distributed/__init__.py:
```diff
@@ -56,33 +56,35 @@ def is_available() -> bool:
         _make_nccl_premul_sum,
     )
 
+    class DistributedPdb(pdb.Pdb):
```
do we want to also make this a public API or we want `breakpoint` be the single public API?

wconstab replied:
not sure, i left it as public thinking that someone might want to further customize. but it is not critical, and we could just make it private for now, which is probably a defensive choice

wanchaol reviewed (approved):
lgtm, thanks for fixing this!! have only one question

wconstab committed:
Update on "Fix torch.distributed.breakpoint"


Switches from calling breakpoint() internally to using a subclass of
Pdb.

Fixes #115685

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

fegin reviewed (approved):
I added the not user facing tag since this is an internal change.

wconstab commented:
@pytorchbot merge

stas00 commented:
@wconstab, thank you for fixing. It's almost perfect now.

I think you need do tell the user to do `up;;n` to have the equivalent of `import pdb; pdb.set_trace()`

with just `up` you end up at `dist.breakpoint()` (test with `w`) and so if you `s` you will step into an unintended frame.

Is there a way to feed `up;;n` automatically to pdb - I guess not since `set_trace` will stop the execution. Hmm, prefilling stdin perhaps?

wconstab commented:
Actually I spent some time trying to fix this, there is a frame argument to set_trace but if I got the parent frame and passed it in it still brakes the same place. I may have been wishfully interpreting what that argument does. 

Writing to stdin seems worth a try.

Note that there is already supposed to be a prompt printed out which explains the 'up' situation.

stas00 commented:
> Note that there is already supposed to be a prompt printed out which explains the 'up' situation.

Yes, what I suggested is to replace `up` with `up;;n` in that suggestion
